### DISCUSSION TITLE
Use XDG Directory Specification for storage and config files.

### DISCUSSION THREAD

**Discussion by masukomi** [pocketbase/pocketbase#1307] [Q&A]
It would seem preferable to store the various `pb_*` directories under the standard XDG locations rather than creating them in the same directory as the executable.

Additionally, by using the XDG spec folks can easily customize where it looks for things just by setting the `XDG_CONFIG_HOME` or `XDG_DATA_HOME` environment variables.

That way we can put `pocketbase` in our `bin` directory without cluttering that up with folders of stuff we wouldn't normally put there.

**Comment by ganigeorgiev** [comment]
Users are expected to add `pb_public` and `pb_migrations` in their repository. Only `pb_data` usually should be gitignored and eventually be moved somewhere else.
The binary is also OK to be added in the repository (this way everyone who clone your repo can start without installing anything).

Additionally it is easier, at least for me, when I know where the `pb_data` directory is located since I can just backup/restore it without trying to figure out the correct path to it (especially when working with different env and os from the one where your app is deployed).

It is possible to specify a different `pb_data`, `pb_public` and `pb_migrations` directories with the `--dir`, `--publicDir` and `--migrationsDir` flags (you can list all flags with `--help`).

**Reply by masukomi** [reply]
maybe this is is my newness to the project but this seems to assume I'd want the binary in my repo, which seems ... atypical. 

Except when working on the source code to a binary, all my binaries live in a bin directory. I think that's kinda normal for people using the binary not developing it.  I don't keep the `git` binary in the git repos that use it. Why would I keep the `pocketbase` binary in the repos that use it? 

I agree that yes it'd make it easier to get up and running on a cloned repo, but only if the person cloning it is running the same OS and chipset. Otherwise it just introduces problems because now they have a binary that they're told will work by the readme, but will actually just be problematic for them. This is the same reason we don't copy the postgres  binary or go binary into our repos.  It'd be great (ignoring space & bandwidth) if that worked but in practice it just doesn't. 

> Additionally it is easier, at least for me, when I know where the pb_data directory is located since I can just backup/restore it without trying to figure out the correct path to it

That makes sense, but I'm not sure how relevant it is. When an app uses XDG I know exactly where its files will be, without it telling me and without having to look, because it's a common and easy standard, and its location doesn't effect my ability to easily back it up or restore it. As soon as you know that `XDG_DATA_HOME` defaults to `~/.local/share` then you know that any app that uses it, has its data stored under `~/.local/share/<app_name>` 

> Only pb_data usually should be gitignored and eventually be moved somewhere else.

Makes sense, but it also sounds like the XDG spec would be perfect for placing / locating the `pb_data` dir. Maybe not all the directories should be there. I think you're right about migrations although i can see an argument both ways. But if `pb_data` lived under `XDG_DATA_HOME` then it would be a consistent "somewhere else" that could be documented, and where users would be able to easily find it on every system because it's always the same place on every OS unless they've gone out of their way to change it. Also, putting it there means folks don't have to worry about remembering to `.gitignore` it. 

> It is possible to specify a different pb_data, pb_public and pb_migrations directories with the --dir, --publicDir and --migrationsDir flags

That's a nice workaround to have available, but it means I've either got to define an alias or type a lot of stuff every time I launch it. Adhering to a default standard that can be tweaked with ENV vars seems a much more elegant and easy solution. Even if that standard isn't XDG.

**Reply by ganigeorgiev** [reply]
> Why would I keep the pocketbase binary in the repos that use it?

Different PocketBase version may come with system migrations that will not work if the binary is started with unsupported pb_data structure. Additionally if you have multiple projects, it is a lot easier, at least for me, if the binary is inside the project folder, especially if the projects are using different PocketBase versions (_there were also suggestions in the past to publish the binary as npm bin module similar to how `sass`, `esbuild`, etc. are distributed, but it will add one more step to the installation instructions and may confuse some users and for now I left this out of the scope_).

> because it's a common and easy standard, and its location doesn't effect my ability to easily back it up or restore it. 

I disagree with this. It "hides" where your data live and "burden" the portability of the application, aka. you can't just zip your app and move it to another server. Currently all users are aware where their `pb_data` is. I doubt that this will be the case if we default to the XDG standard (even if it is documented). It will also cause confusion and issues if multiple PocketBase applications are running on the same server (especially if they are using different versions).

There are no plans for now to change the default location of the directory, neither replacing the cli flags with env variables (_this will not work with multiple applications on the same server_).

The only thing related to the directories I have in my todo for now is to eventually consider wrapping all pb related directories in one (and adding a default .gitignore when creating `pb_data`), because in the future we may add support for optional js `pb_hooks` and it may become too verbose/cluttered to have 4 different dir flags.
